### PR TITLE
Pybind refactor

### PR DESCRIPTION
**Pull request by vlarsson**
Contains pybind refactorization and alignment fixes.

Bumped version to 2.0.1.

### CODE REVIEW

**Comment by pablospe** [comment]
What about having a Version.txt file? Suggested in this comment: https://github.com/vlarsson/PoseLib/pull/58#issuecomment-1427062354

File: PoseLib/misc/essential.cc [hunk 1]
```diff
@@ -77,7 +77,30 @@ bool check_cheirality(const CameraPose &pose, const Eigen::Vector3d &p1, const E
     return lambda1 > min_depth && lambda2 > min_depth;
 }
 
-void motion_from_essential(const Eigen::Matrix3d &E, const Eigen::Vector3d &x1, const Eigen::Vector3d &x2,
+
+// wrappers for vectors
+bool check_cheirality(const CameraPose &pose, const std::vector<Eigen::Vector3d> &x1, const std::vector<Eigen::Vector3d> &x2,
+                      double min_depth) {
+    for(size_t i = 0; i < x1.size(); ++i) {
```
Comment (pablospe) [inline PoseLib/misc/essential.cc:84]: maybe checking that x1 and x2 have the same size?

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            self.build_extension(ext)
+
+    def build_extension(self, ext):
+        extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
+        cmake_args = ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + extdir,
+                      '-DPYTHON_EXECUTABLE=' + sys.executable,
+                      '-DPython_EXECUTABLE=' + sys.executable,
+                      '-DPYTHON_PACKAGE=ON']
+
+        cfg = 'Debug' if self.debug else 'Release'
```
Comment (pablospe) [inline setup.py:41]: Make sure default is Release.

File: pybind/CMakeLists.txt [hunk 3]
```diff
@@ -1,35 +1,12 @@
...
-            -DPYTHON_PACKAGE=ON
-            -DPYTHON_PACKAGE_DST_DIR=${PYTHON_PACKAGE_DST_DIR}
-            -P ${CMAKE_CURRENT_SOURCE_DIR}/make_install_pip_package.cmake
-    DEPENDS pip-package
-)
+pybind11_add_module(pyposelib MODULE pyposelib.cc)
+target_link_libraries(pyposelib PUBLIC PoseLib::PoseLib Eigen3::Eigen)
+target_compile_features(pyposelib PRIVATE cxx_std_17)
+set_target_properties(pyposelib PROPERTIES OUTPUT_NAME poselib)
+target_compile_options(pyposelib PRIVATE -march=native -Wall -Werror -fPIC -Wno-ignored-optimization-argument)
```
Comment (pablospe) [inline pybind/CMakeLists.txt:12]: GitHub recommends adding a newline at the end of files to comply with POSIX standards.

File: README.md
Comment (pablospe) [inline README.md:1]: Search for `version.h`

Comment (pablospe) [reply README.md:1]: This needs be updated I think?

File: CMakeLists.txt [hunk 4]
```diff
@@ -29,10 +31,9 @@ endif()
 if(MSVC)
 	target_compile_options(${LIBRARY_NAME} PRIVATE /bigobj /fp:fast)
 else()
+	# If you change this, make sure to update the corresponding line in the pybind CMakeLists
```
Comment (pablospe) [inline CMakeLists.txt:34]: Could this be centralized somehow?

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,12 +1,14 @@
...
 
-project(PoseLib VERSION 2.0.0)
+project(PoseLib VERSION 2.0.1)
 
 # Set variables
 set(LIBRARY_NAME   ${PROJECT_NAME})
 set(LIBRARY_FOLDER ${PROJECT_NAME})
 include(${PROJECT_SOURCE_DIR}/cmake/SetEnv.cmake)
 
+
```
Comment (pablospe) [inline CMakeLists.txt:10]: nit: white lines

File: CMakeLists.txt (hunk 4 above)
Comment (vlarsson) [reply CMakeLists.txt:34]: Yes, I have changed this in another WIP branch. See here 
https://github.com/vlarsson/PoseLib/blob/camera_models/CMakeLists.txt#L10-L18

File: README.md
Comment (vlarsson) [reply README.md:1]: I am not sure what you mean. There is still a version.h.in that is setup by cmake.

File: setup.py (hunk 2 above)
Comment (vlarsson) [reply setup.py:41]: This defaults to release no? l41 ?

File: CMakeLists.txt (hunk 5 above)
Comment (vlarsson) [reply CMakeLists.txt]: Fixed in 1f94afa

File: pybind/CMakeLists.txt (hunk 3 above)
Comment (vlarsson) [reply pybind/CMakeLists.txt]: Fixed in 1f94afa

File: setup.py (hunk 2 above)
Comment (pablospe) [reply setup.py:41]: Didn't read properly, sorry. Marked as resolved

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        env = os.environ.copy()
+        env['CXXFLAGS'] = '{} -DVERSION_INFO=\\"{}\\"'.format(
+            env.get('CXXFLAGS', ''),
+            self.distribution.get_version()
+        )
+        if not os.path.exists(self.build_temp):
+            os.makedirs(self.build_temp)
+        print(['cmake', ext.sourcedir] + cmake_args)
+        print(build_args)
+        subprocess.check_call(['cmake', ext.sourcedir] + cmake_args, cwd=self.build_temp, env=env)
```
Comment (pablospe) [inline setup.py:68]: Where is this being built? Shouldn't it be something like in the README.md `_build`?

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        subprocess.check_call(['cmake', '--build', '.'] + build_args, cwd=self.build_temp)
+
+
+
+# The information here can also be placed in setup.cfg - better separation of
+# logic and declaration, and simpler if you include description/version in a file.
+setup(
+    name="poselib",
+    version="2.0.1",
```
Comment (pablospe) [inline setup.py:77]: What about having a Version.txt file? Suggested in this comment: https://github.com/vlarsson/PoseLib/pull/58#issuecomment-1427062354

File: README.md
Comment (pablospe) [reply README.md:1]: Ok, yes, I got confused because it was removed from `pybind/helpers.h`.

**Review by pablospe — APPROVED** [review]

File: setup.py (hunk 7 above)
Comment (vlarsson) [reply setup.py:77]: Yes I think this would be good. My cmake-magic is too weak to implement it though :)

File: setup.py (hunk 6 above)
Comment (vlarsson) [reply setup.py:68]: As far as I know it's build in /tmp/.

**Pull request merged**